### DISCUSSION TITLE
General planning and content categoriztion

### DISCUSSION THREAD

**Discussion by pnpancholi** [PradhumnaPancholi/solidity-cookbook#4] [Ideas]
Hey @nonsensecodes, the project has been set up and I have done some cleanup. I am thinking as following:

1. Introduction - Intent, ideas, expectations, requirements,  how-to(s), and maybe a quick story behind why it exists.
2. Basics - This section will cover functions, events, variables, modifiers, etc. This section will have a part (last) where we set up the local development environment, write a simple contract, write tests for it, and deploy on Kovan.
3. Intermediate (looking for a better name) -  This is where we will start writing a bit more complex contracts. Exposing concepts like interface (connect to other contracts), connecting to oracles, getting comfortable with relatively larger contracts(spread across multiple files), upgradable contracts, gas optimization, security testing( both unit testing and introducing to some static testing tool), etc
4. Advanced (much more will be added to this section) - Writing oracles, zero-knowledge -proofs (implementation),  exposing tools like OpenZeppelinn Defender (and other tools in the suite), discussing popular hacks and solutions to already discovered vulnerabilities, etc.

This is not some sort of finalized curriculum, just some idea to start working on and we will always be updating the material.

Let me know your thoughts on this.

**Comment by lvnacy** [comment]
This is a really great outline. Step by step, the complexity increases appropriately as you move up in levels. I think with the Solidity Learning group, we can help flesh out much of the basics and some, if not all, of the intermediate level. How do you want us to approach contributions? I’m happy to help with any writing that needs to be done.
